### PR TITLE
fix(lambda-python-alpha): use function architecture (#18696)

### LINKED ISSUE #18696: (lambda-python): arm64 architecture is not respected
### What is the problem?

Hey,

Not quite sure if that's a problem with lambda or with codepipelines or at all with cdk.
Been trying to understand it but I just couldn't. Here's what is happening.

I have a `pipelines.CodePipeline` that is configured with `code_build_defaults` that states that I wish to run builds with `LinuxBuildImage.AMAZON_LINUX_2_ARM_2`. Obviously later on my lambda functions are configured as follow:

```
PythonFunction(..., runtime=Runtime.PYTHON_3_9, architecture=Architecture.ARM_64)
```

That being said I expect that when `CodeBuilds` runs my `synth` step it will run on top of `arm64` and lambda will be build using the same architecture. However in the logs of job I see this:

```sh
111 | Status: Downloaded newer image for public.ecr.aws/sam/build-python3.9:latest
112 | ---> 7925e2bf1015
113 | Step 3/8 : ARG PIP_INDEX_URL
114 | ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
115 | ---> Running in 47078dfc29f5
116 | Removing intermediate container 47078dfc29f5
117 | ---> b95a315c94b2
118 | Step 4/8 : ARG PIP_EXTRA_INDEX_URL
119 | ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
120 | ---> Running in 684bb68c4f18
121 | Removing intermediate container 684bb68c4f18
122 | ---> 175da3dd341e
123 | Step 5/8 : ARG HTTPS_PROXY
124 | ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
125 | ---> Running in 15976d15e0ae
126 | Removing intermediate container 15976d15e0ae
127 | ---> 28715ae8a9d6
128 | Step 6/8 : RUN pip install --upgrade pip
129 | ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
130 | ---> Running in 515be8ae3321
131 | standard_init_linux.go:228: exec user process caused: exec format error
132 | The command '/bin/sh -c pip install --upgrade pip' returned a non-zero code: 1
133 | jsii.errors.JavaScriptError:
```

Which seems to contradict to my desire setup of `arm64`. I expected `public.ecr.aws/sam/build-python3.9:latest-arm64` would be downloaded instead of `public.ecr.aws/sam/build-python3.9:latest` (lack or **arm64**) in the end.

That happens with `cdk 2.8.0`.
Last successful build I managed to execute with here was done with `cdk 1.134.0`.

PS. My stack defines also `AwsCustomResoure` which is another lambda deployment. Might it be that arm64 and custom resource lambda conflicts each other?

### Reproduction Steps

```python3
import typing as t

import aws_cdk as cdk
from aws_cdk import (
    aws_codebuild as cb,
    aws_codepipeline as cp,
    aws_iam as iam,
    aws_lambda as fn,
    aws_lambda_python as fn_python,
    pipelines,
)
from constructs import Construct

class Pipeline(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        **kwargs: t.Any,
    ) -> None:
        super().__init__(
            scope,
            construct_id,
            **kwargs,
        )

        connection_arn =  '' # some connection arn here
        repo_name = '' # some repo name here

        pipeline = pipelines.CodePipeline(
            self,
            'Pipeline',
            pipeline_name='IAMSetupPipeline',
            synth=pipelines.ShellStep(
                'Synth',
                input=pipelines.CodePipelineSource.connection(
                    repo_name
                    'master',
                    connection_arn=connection_arn,
                ),
                install_commands=[
                    'npm install -g aws-cdk@latest',
                    'pip install -r requirements.txt',
                ],
                commands=['cdk synth pipeline'],
            ),
            code_build_defaults=pipelines.CodeBuildOptions(
                build_environment=cb.BuildEnvironment(
                    build_image=cb.LinuxBuildImage.AMAZON_LINUX_2_ARM_2,
                ),
            ),
            self_mutation=True,
            docker_enabled_for_synth=True,
        )

        # stages
        pipeline.add_stage(
            AppStage(
                self,
               'bug',
                env=cdk.Environment(
                    account='000000000000',
                    region='eu-central-1',
                ),
            ),
        )

        pipeline.build_pipeline()


class AppStage(cdk.Stage):
    def __init__(
        self,
        scope: Construct,
        stage_id: str,
        **kwargs: t.Any,
    ) -> None:
        super().__init__(scope, stage_id, **kwargs)
        LambdaStack(self, 'stack')

class LambdaStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        **kwargs: t.Any,
    ) -> None:
        super().__init__(
            scope,
            construct_id,
            **kwargs,
        )
 
        fn_python.PythonFunction(
            self,
            'ARM',
            entry='/your/entry',
            runtime=fn.Runtime.PYTHON_3_9,
            architecture=fn.Architecture.ARM_64,
        )

app = cdk.App()
Pipeline(
    app, 
    'Pipeline', 
    env=cdk.Environment(
                    account='000000000000',
                    region='eu-central-1',
))
app.synth()
``` 

CODE is not exactly to be run as-is. It lacks a python lambda code (this is what I have been using) and it lacks proper account and region

### What did you expect to happen?

Stack is built correctly inside of deployed pipeline and ARM64 is used.

### What actually happened?

Stack fails to built with code linked in first section.

### CDK CLI Version

2.8.0

### Framework Version

_No response_

### Node.js Version

14.17.5

### OS

MacOs BigSur

### Language

Python

### Language Version

3.10.1

### Other information

If I messed something up. I am sorry :(

### PR DESCRIPTION
**Pull request by lucacucchetti**
With this change, architecture when bundling is inferred from the target architecture of the Lambda function.

Closes #18696.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by lucacucchetti** [comment]
Hello, I tried adding an integration test but perhaps predictably bundling failed with `exec format error`:

```
 > [2/2] RUN     python -m venv /usr/app/venv &&     mkdir /tmp/pip-cache &&     chmod -R 777 /tmp/pip-cache &&     pip install --upgrade pip &&     mkdir /tmp/poetry-cache &&     chmod -R 777 /tmp/poetry-cache &&     pip install pipenv==2022.4.8 poetry==1.5.1 &&     rm -rf /tmp/pip-cache/* /tmp/poetry-cache/*:
1.204 exec /bin/sh: exec format error
------
Dockerfile:20

...

Command: docker build -t cdk-e0fb70af5dd7d3ee8ee52bbcc32ec617017cebac8121a5ccbc396830b855b75f --platform "linux/arm64" --build-arg "IMAGE=public.ecr.aws/sam/build-python3.11" "/workspace/aws-cdk/packages/@aws-cdk/aws-lambda-python-alpha/lib"
    at dockerExec (/workspace/aws-cdk/packages/aws-cdk-lib/core/lib/private/asset-staging.js:187:15)
```

I tested this in a GitPod environment (from the [contribution guide](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#gitpod)) which I thought would closely resemble the automated PR Linter:

```
uname -m
x86_64
```

Please let me know if I should still pursue making an integ test for this situation.

**Comment by lucacucchetti** [comment]
Sorry about that, since I mentioned this on [this issue](https://github.com/aws/aws-cdk/issues/18696) a while back and saw no activity assumed no one was working on it.

@kaizencc do we need to get the `pr-linter/exempt-integ-test` label in for the Linter to pass?

**Comment by Zordrak** [comment]
I would have loved some clear documentation here. This is a major pipeline breaking change for pipelines that are developing, building and testing lambdas in x86_64 environments, but deploying the lambdas to ARM.

I am now trying to find a way to reproduce the original behaviour, but I am struggling for a viable solution.

**Comment by lucacucchetti** [comment]
I sympathise, in the original issue I [commented something](https://github.com/aws/aws-cdk/issues/18696#issuecomment-1798241388) to that effect:

> I think the contract should change, architecture should be an attribute of BundlingOptions (since it is possible in some situations to want to target one architecture while building in another), regardless, the current behaviour is not what the docs describe and so I would call it a bug.

The [documented](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-lambda-python-alpha-readme.html) behaviour does say "platform based on the target architecture of the Lambda function".

For your case, can you do the following?

```
super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      bundling: {
          architecture: Architecture. X86_64,
      } as any,
      architecture: Architecture.ARM_64,
      ...
```

If you check the diff, `bundling` overrides the default of the lambda architecture.

**Comment by jbschooley** [comment]
This change broke my pipeline as well. I'm using this for layers:

```
bundling: {
    image: cdk.DockerImage.fromRegistry('public.ecr.aws/sam/build-python3.11:latest-arm64'),
}
```

to force it to use the arm64 image to build.

If I take that out and add `architecture: Architecture.X86_64` it then builds with the x86 image, installing the x86 builds of dependencies.

If I set `architecture: Architecture.ARM64` then it passes in `--platform "linux/arm64"` but tries to build with the x86 image and fails.

Both `architecture: Architecture.ARM64` and setting the arm64 image, it uses the image but does not pass in `--platform "linux/arm64"` and fails with

```
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64/v3) and no specific platform was requested
exec /usr/bin/bash: exec format error
```

adding `platform: 'linux/arm64'` to bundling does nothing.

Basically, my arm64 build seems to be broken and I have no way to update my layers without moving them, and all the functions they are used in, to x86.

**Comment by lucacucchetti** [comment]
Just to confirm, you are talking about a problem while building layers and not lambdas? 

This change did not update any layer related code or behaviour.

**Comment by jbschooley** [comment]
Correct. I don't use this to build functions, I keep all my dependencies in layers. If this change didn't update layer related code then I'm not sure what happened because my layer building pipeline is broken in the same way.
